proposal/Luke2336/README.rst

```diff
@@ -0,0 +1,43 @@
+===================
+Graph Visualization
+===================
+
+
+Basic Information
+=================
+
+- GitHub repository: https://github.com/Luke2336/graph-visualization
+
+Problem to Solve
+================
+
+Visualizing a graph can help people study graph theory. In this project, I would like to use force-directed method (Fruchterman-Reingold Algorithm) to layout a simple undirected graph to let the number of edge intersections as small as possible.
+
+
+Prospective Users
+=================
+
+The prospective users are poeple who want to visualize a graph.
+
+System Architecture
+===================
+
+- Input format:
+	The first line contains two non-negative integers n and m.
+	For 2 ~ m + 1 lines, there are two integer a and b, which means there is an edge between vertex a and b.
+
+
+API Description
+===============
+
+
+Engineering Infrastructure
+==========================
+
```

Comment:
Verification is particularly difficult for anything involving image (for which visualization incurs).  Do you have an idea how to write good tests for it?  (Reminder: it is HARD.)

Good tests should be easy to fail with wrong behavior, but resilient to good changes to the system.

Reply:
In my opinion, the main idea of this project is to layout a graph on a 2D plane. Can my test plan only test the geometric points and segments generating by my algorithm? For example, I can check whether there are segments overlapping on a line or some vertices being too close.

I realize that the verification of visualization is really hard since I have to scan an image and transfer it into geometric points and segments. In short, I think the scale of this step is strongly larger than my original project.

Reply:
> In my opinion, the main idea of this project is to layout a graph on a 2D plane. 

In other words, the essence is to transform a graph to a (vector) "graphic".  In this case, it is not necessarily visualization.

Converting a graph to a vector graphic sounds much generic, and the problem becomes simpler, simultaneously.

> Can my test plan only test the geometric points and segments generating by my algorithm?

By casting the problem from visualization to conversion to graphics, yes you can then test against the geometry.

> For example, I can check whether there are segments overlapping on a line or some vertices being too close.

Yes you can, but how are you going to check for overlapping of two line segments?  The check for overlapping should incur significant runtime.

> 
> I realize that the verification of visualization is really hard since I have to scan an image and transfer it into geometric points and segments. In short, I think the scale of this step is strongly larger than my original project.

The analysis makes sense.  Let's keep discussing it.

Reply:
In the updated version, this point has not been addressed.